### PR TITLE
Refine evaluation code.

### PR DESCRIPTION
**Pull request by qingqing01**

### CODE REVIEW

File: fluid/face_detection/reader.py [hunk 1]
```diff
@@ -331,8 +308,9 @@ def test(settings, file_list):
 
     def reader():
         for index_image in file_dict.keys():
-            image_name = file_dict[index_image][0] + '.jpg'
+            image_name = file_dict[index_image][0]
             image_path = os.path.join(settings.data_dir, image_name)
+            print(image_path)
```
Comment (baiyfbupt) [inline fluid/face_detection/reader.py]: is this line still needed?

File: fluid/face_detection/widerface_eval.py [hunk 2]
```diff
@@ -4,68 +4,132 @@
...
 
-def write_to_txt(image_path, f, nms_out):
+        save_widerface_bboxes(image_path, dets, pred_dir)
+
+    print("Finish evaluation.")
+
+
+def save_widerface_bboxes(image_path, bboxes_scores, output_dir):
+    """
+    Save predicted results, including bbox and score into text file.
```
Comment (baiyfbupt) [inline fluid/face_detection/widerface_eval.py]: `bbox`->`bboxes`,`score`->`scores`

File: fluid/face_detection/widerface_eval.py [hunk 3]
```diff
@@ -4,68 +4,132 @@
...
+        print("No face detected")
+        return np.array([[0, 0, 0, 0, 0]])
+    det_conf = detection[:, 1]
+    det_xmin = image_shape[2] * detection[:, 2] / shrink
+    det_ymin = image_shape[1] * detection[:, 3] / shrink
+    det_xmax = image_shape[2] * detection[:, 4] / shrink
+    det_ymax = image_shape[1] * detection[:, 5] / shrink
+
+    det = np.column_stack((det_xmin, det_ymin, det_xmax, det_ymax, det_conf))
+    keep_index = np.where(det[:, 4] >= 0)[0]
```
Comment (baiyfbupt) [inline fluid/face_detection/widerface_eval.py]: `fluid.layers.detection_output ` return bboxes `scores>score_threshold`, is 130-131 line still needed?

File: fluid/face_detection/widerface_eval.py [hunk 4]
```diff
@@ -197,31 +189,32 @@ def flip_test(image, shrink):
...
         det_b = np.row_stack((det_b, detect_face(image, max_shrink)))
 
-    # enlarge only detect small face
+    # Enlarged images are only used to detect small face.
     if bt > 1:
         index = np.where(
             np.minimum(det_b[:, 2] - det_b[:, 0] + 1,
                        det_b[:, 3] - det_b[:, 1] + 1) < 100)[0]
         det_b = det_b[index, :]
+    # Reduced images are only used to detect big face.
```
Comment (baiyfbupt) [inline fluid/face_detection/widerface_eval.py]: `Reduced`->`Shrinked`? same as line 192

File: fluid/face_detection/widerface_eval.py [hunk 5]
```diff
@@ -242,12 +235,13 @@ def multi_scale_test_pyramid(image, max_shrink):
...
         if (st[i] <= max_shrink):
             det_temp = detect_face(image, st[i])
-            # enlarge only detect small face
+            # Enlarged images are only used to detect small face.
             if st[i] > 1:
                 index = np.where(
                     np.minimum(det_temp[:, 2] - det_temp[:, 0] + 1,
                                det_temp[:, 3] - det_temp[:, 1] + 1) < 100)[0]
                 det_temp = det_temp[index, :]
+            # Reduced images are only used to detect big face.
```
Comment (baiyfbupt) [inline fluid/face_detection/widerface_eval.py]: Shrinked?  `faces`

File: fluid/face_detection/widerface_eval.py [hunk 6]
```diff
@@ -197,31 +189,32 @@ def flip_test(image, shrink):
...
         np.maximum(det_s[:, 2] - det_s[:, 0] + 1, det_s[:, 3] - det_s[:, 1] + 1)
         > 30)[0]
     det_s = det_s[index, :]
-    # enlarge one times
+    # Enlarge one times
     bt = min(2, max_shrink) if max_shrink > 1 else (st + max_shrink) / 2
     det_b = detect_face(image, bt)
 
-    # enlarge small image x times for small face
+    # Enlarge small image x times for small face
```
Comment (baiyfbupt) [inline fluid/face_detection/widerface_eval.py]: `faces`

File: fluid/face_detection/widerface_eval.py [hunk 7]
```diff
@@ -197,31 +189,32 @@ def flip_test(image, shrink):
...
     if max_shrink > 2:
         bt *= 2
         while bt < max_shrink:
             det_b = np.row_stack((det_b, detect_face(image, bt)))
             bt *= 2
         det_b = np.row_stack((det_b, detect_face(image, max_shrink)))
 
-    # enlarge only detect small face
+    # Enlarged images are only used to detect small face.
```
Comment (baiyfbupt) [inline fluid/face_detection/widerface_eval.py]: `faces`

File: fluid/face_detection/widerface_eval.py (hunk 4 above)
Comment (baiyfbupt) [reply fluid/face_detection/widerface_eval.py]: `faces`

File: fluid/face_detection/reader.py (hunk 1 above)
Comment (qingqing01) [reply fluid/face_detection/reader.py]: Removed.

File: fluid/face_detection/widerface_eval.py [hunk 8]
```diff
@@ -242,12 +235,13 @@ def multi_scale_test_pyramid(image, max_shrink):
     for i in range(len(st)):
         if (st[i] <= max_shrink):
             det_temp = detect_face(image, st[i])
-            # enlarge only detect small face
+            # Enlarged images are only used to detect small face.
```
Comment (baiyfbupt) [inline fluid/face_detection/widerface_eval.py]: `faces`

File: fluid/face_detection/widerface_eval.py (hunk 3 above)
Comment (qingqing01) [reply fluid/face_detection/widerface_eval.py]: Yeah, removed.

File: fluid/face_detection/widerface_eval.py (hunk 6 above)
Comment (qingqing01) [reply fluid/face_detection/widerface_eval.py]: Done.

File: fluid/face_detection/widerface_eval.py (hunk 7 above)
Comment (qingqing01) [reply fluid/face_detection/widerface_eval.py]: Done.

File: fluid/face_detection/widerface_eval.py (hunk 4 above)
Comment (qingqing01) [reply fluid/face_detection/widerface_eval.py]: Done.

File: fluid/face_detection/widerface_eval.py (hunk 5 above)
Comment (qingqing01) [reply fluid/face_detection/widerface_eval.py]: Done.

File: fluid/face_detection/widerface_eval.py (hunk 8 above)
Comment (qingqing01) [reply fluid/face_detection/widerface_eval.py]: Done.

**Comment by qingqing01** [comment]
@baiyfbupt Thanks!

**Review by baiyfbupt — APPROVED** [review]
LGTM

**Pull request merged**